# Segmenting and Clustering Neighborhoods in Toronto

## <u> Task One: A data frame for Toronto neighborhoods </u> 

#### importing the required libraries

In [14]:
# First updating the libraries as required (code commented for earlier installs)

#!pip install -U numpy

#!pip install -U pandas

#!pip install -U scipy==1.4.1

#!pip install -U scikit-learn

#!pip install -U imbalanced-learn
print("libs updated")

libs updated


In [2]:
#import pandas library
import pandas as pd # library for data analysis and manipulation. will convert data to DataFrame objects 

# import requests
import requests # library to handle web requests

# import BeautifulSoup
from bs4 import BeautifulSoup  # library for pulling data out of HTML and XML files

####  defining the wikipedia data source and creating a Beautifulsoup object

In [15]:
# defining and retreiving the URL and creating a Beautifulsoup object 
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

tn_data = requests.get(url).text

soup = BeautifulSoup(tn_data)

# print(soup.prettify()) # displaying the Beautifulsoup object contents


#### extracting table data from the soup object (the nested data structure document above)

In [4]:
# Creating an empty list to hold table data 
table_data=[]

# finding the table - and table data
table=soup.find('table')

# creating a dictionary called t_Cols having 3 keys PostalCode, Borough and Neighborhood

# ...and ignoring cells with a borough that is **Not assigned**
for row in table.findAll('td'):
    t_Cols = {}
    if row.span.text=='Not assigned':           # ignoring cells with a borough that is **Not assigned**
        pass
    else:
        t_Cols['PostalCode'] = row.p.text[:3]
        t_Cols['Borough'] = (row.span.text).split('(')[0]
        t_Cols['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_data.append(t_Cols)  # appending the dictionary data to the list object

# print(table_data)    # viewing the table data obtained


# converting the list object into a pandas data frame object
df=pd.DataFrame(table_data)

# cleaning Borough Names
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [5]:
# The data frame
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### Number of rows

In [6]:
no=df.shape[0]
print("The resultant dataframe has {} rows".format(no))


The resultant dataframe has 103 rows


## <u> Task Two: getting the latitude and the longitude coordinates of each neighborhood </u>

##### installing amd importing the required modules and libraries

In [7]:
# The Canada postal codes which have alphabets are not easily parsed in the Geocoder library 
# and we use the pgeocode module to fetch the latitude and longitude values

!pip install pgeocode
import pgeocode

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes


##### fetching the geographical coordinates of a given postal code with pgeocode using a looping function untill cordinates for each postal code are found

In [8]:
#pgeocode.Nominatim('ca')

# defining a pgeocode instance for canada addresses
geolocator = pgeocode.Nominatim('ca')

# defining the list of postal addresses from our main dataframe
postal_codes = df['PostalCode'].tolist()

# initializing empty lists to store the latitudes and longitudes values
latitudes = []
longitudes = []

# looping through each postal code to get the addresses
for i, postal_code in enumerate(postal_codes):
    
    # Running a geo-localisation query for each Postal Code 
    g = geolocator.query_postal_code(postal_code)
    
    if not g.empty:
        # storing the latitude and longitudes values from the query to the lists created earlier 
        latitudes.append(g.latitude)
        longitudes.append(g.longitude)    

##### converting the pair of latitudes and longitudes lists into a pandas dataframe for confirmation 

In [9]:
lists = [latitudes, longitudes]
df_coords = pd.concat([pd.Series(x) for x in lists], axis=1)
df_coords.columns=['latitudes', 'longitudes'] 
df_coords.head(2)

latitudes  longitudes
0    43.7545    -79.3300
1    43.7276    -79.3148

##### attaching the two as new columns into our main dataframe of  Toronto neighborhoods

##### This is the final dataframe for task two

In [10]:
df['Latitude']=df_coords['latitudes']
df['Longitude']=df_coords['longitudes']
pd.options.display.max_rows = None
print(df.shape)
df

(103, 5)


PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

# <u> Task 3: Exploring and clustering the neighborhoods in Toronto </u>

##### 3.1 Here we will be working with only the boroughs whose names containing the word 'Toronto'

In [11]:
# filtering out the boroughs with the name "Toronto" into a new dataframe - df_viz
df_viz =df[df['Borough'].str.contains('Toronto')]

# reseting the index for the new dataframe
df_viz.reset_index(inplace=True, drop=True)

# size of df_viz 
print(df_viz.shape)
df_viz

(39, 5)


PostalCode                 Borough  \
0         M5A        Downtown Toronto   
1         M5B        Downtown Toronto   
2         M5C        Downtown Toronto   
3         M4E            East Toronto   
4         M5E        Downtown Toronto   
5         M5G        Downtown Toronto   
6         M6G        Downtown Toronto   
7         M5H        Downtown Toronto   
8         M6H            West Toronto   
9         M4J  East York/East Toronto   
10        M5J        Downtown Toronto   
11        M6J            West Toronto   
12        M4K            East Toronto   
13        M5K        Downtown Toronto   
14        M6K            West Toronto   
15        M4L            East Toronto   
16        M5L        Downtown Toronto   
17        M4M            East Toronto   
18        M4N         Central Toronto   
19        M5N         Central Toronto   
20        M4P         Central Toronto   
21        M5P         Central Toronto   
22        M6P            West Toronto   
23        M4R         Central Toronto   
24        M5R         Central Toronto   
25        M6R            West Toronto   
26        M4S         Central Toronto   
27        M5S        Downtown Toronto   
28        M6S            West Toronto   
29        M4T         Central Toronto   
30        M5T        Downtown Toronto   
31        M4V         Central Toronto   
32        M5V        Downtown Toronto   
33        M4W        Downtown Toronto   
34        M5W  Downtown Toronto Stn A   
35        M4X        Downtown Toronto   
36        M5X        Downtown Toronto   
37        M4Y        Downtown Toronto   
38        M7Y   East Toronto Business   

                                         Neighborhood  Latitude  Longitude  
0                           Regent Park, Harbourfront   43.6555   -79.3626  
1                            Garden District, Ryerson   43.6572   -79.3783  
2                                      St. James Town   43.6513   -79.3756  
3                                         The Beaches   43.6784   -79.2941  
4                                         Berczy Park   43.6456   -79.3754  
5                                  Central Bay Street   43.6564   -79.3860  
6                                            Christie   43.6683   -79.4205  
7                            Richmond, Adelaide, King   43.6496   -79.3833  
8                        Dufferin, Dovercourt Village   43.6655   -79.4378  
9                                  The Danforth  East   43.6872   -79.3368  
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936  
11                           Little Portugal, Trinity   43.6480   -79.4177  
12                       The Danforth West, Riverdale   43.6803   -79.3538  
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823  
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301  
15                     India Bazaar, The Beaches West   43.6693   -79.3155  
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823  
17                                    Studio District   43.6561   -79.3406  
18                                      Lawrence Park   43.7301   -79.3935  
19                                           Roselawn   43.7113   -79.4195  
20                                   Davisville North   43.7135   -79.3887  
21                           Forest Hill North & West   43.6966   -79.4120  
22                      High Park, The Junction South   43.6605   -79.4633  
23                                 North Toronto West   43.7143   -79.4065  
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035  
25                             Parkdale, Roncesvalles   43.6469   -79.4521  
26                                         Davisville   43.7020   -79.3853  
27                     University of Toronto, Harbord   43.6629   -79.3987  
28                                 Runnymede, Swansea   43.6512   -79.4828  
29                        Moore Park, Summerhill Eas

we see that the new dataframe has 39 neighborhoods

##### 3.2 Next, it's segmenting and Clustering Neighborhoods in selected Boroughs

 ###### We start by importing the required libraries

In [12]:
import numpy as np # library to handle data in a vectorized manner
print("imported numpy")

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
print("imported pandas")

import json # library to handle JSON files
print("imported json")

!conda install -c conda-forge geopy --yes # code commented if installed earlier
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
print("installed geopy")

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
print("imported requests")

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
print("imported matplotlib")

# import k-means for clustering stage
from sklearn.cluster import KMeans
print("imported Kmeans")

!conda install -c conda-forge folium=0.5.0 --yes # code commented if installed earlier
print("installed folium")

import folium # map rendering library
print("imported folium")

print('Libraries updated/installed. All imported!')

imported numpy
imported pandas
imported json
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.7-main

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |           1_llvm           5 KB  conda-forge
    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge
    _pytorch_select-0.2        |            gpu_0           2 KB
    absl-py-0.12.0             |     pyhd8ed1ab_0          96 KB  conda-forge
    aiohttp-3

##### 3.3 Creating a map of Toronto with neighborhoods superimposed on top.
##### 3.3.1 First we get the latitude and longitude for  Toronto

In [13]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geographical coordinate of Toronto City are 43.6534817, -79.3839347.


##### 3.3.2 Then we create a geo-map visualization of all the neighborhoods selected above (i.e, those with boroughs containing the name 'Toronto')  

In [14]:
# create map of Toronto using latitude and longitude values
neighborhoods = df_viz
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)


# add markers to map for each neighborhood
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  

# display the map    
map_Toronto

Above is the map of Toronto for all neighborhoods with borough names containing the word Toronto

#### 3.4 Further more exploration of the selected neighborhoods
#### 3.4.1 Defining the  Foursquare Credentials and Version to be used in api calls  

In [15]:
# @hidden
CLIENT_ID = 'BML5E42WUO15B0NWXPECBYDTM24OEM1Z5TLTGSXWHL302ORB' #  Foursquare ID
CLIENT_SECRET = 'C2KDHSGGXQ02XG12NWAC1KKMTYPJ55QV2HJ2LOZNEXDYQ52H' #  Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

#print('My credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

##### 3.4.2 Let's explore the first neighborhood in our dataframe.

###### 3.4.2.1  <i> .The neighborhood's name....</i>


In [45]:
#getting the nieghborhood name ar the first row of our data

neighborhoods.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

###### 3.4.2.2. <i> The neighborhood's latitude and longitude values.

In [17]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6555, -79.3626.


###### 3.4.2.3  <i>  Getting the top 100 venues that are in Regent Park & Harbourfront within a radius of 500 meters.

<i>  First, creating the GET request URL. Name of URL is url.


In [18]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# creating URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # displaying the URL

'https://api.foursquare.com/v2/venues/explore?&client_id=BML5E42WUO15B0NWXPECBYDTM24OEM1Z5TLTGSXWHL302ORB&client_secret=C2KDHSGGXQ02XG12NWAC1KKMTYPJ55QV2HJ2LOZNEXDYQ52H&v=20180605&ll=43.6555,-79.3626&radius=500&limit=100'

<i> Second, sending the GET request and examining the resutls

In [19]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60af8064e66f2f772397b8ec'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 21,
  'suggestedBounds': {'ne': {'lat': 43.660000004500006,
    'lng': -79.3563918719477},
   'sw': {'lat': 43.6509999955, 'lng': -79.36880812805231}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '53b8466a498e83df908c3f21',
       'name': 'Tandem Coffee',
       'location': {'address': '368 King St E',
        'crossStreet': 'at Trinity St',
        'lat': 43.65355870959944,
        'lng': -79.36180945913513,
        'labeledLatLngs': [{'label': 'di

<b> We see that all the information is in the items key. </b>
- The information for the first venue  is replicated here:

In [20]:
venue_1 = results['response']['groups'][0]['items'][0]
venue_1

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '53b8466a498e83df908c3f21',
  'name': 'Tandem Coffee',
  'location': {'address': '368 King St E',
   'crossStreet': 'at Trinity St',
   'lat': 43.65355870959944,
   'lng': -79.36180945913513,
   'labeledLatLngs': [{'label': 'display',
     'lat': 43.65355870959944,
     'lng': -79.36180945913513}],
   'distance': 225,
   'cc': 'CA',
   'city': 'Toronto',
   'state': 'ON',
   'country': 'Canada',
   'formattedAddress': ['368 King St E (at Trinity St)',
    'Toronto ON',
    'Canada']},
  'categories': [{'id': '4bf58dd8d48988d1e0931735',
    'name': 'Coffee Shop',
    'pluralName': 'Coffee Shops',
    'shortName': 'Coffee Shop',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-53b8466a498e83

<i> we define the get_category_type function from the Foursquare lab here.

In [21]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

 <i> Cleaning the json and structure into a pandas dataframe.

In [22]:
# defining the venues data path 
venues = results['response']['groups'][0]['items']
 
# converting the venues data to pandas 
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns of interest
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row using the function defined in cell above
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# cleaning column names
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

# displating the venues near 
nearby_venues.head()

name      categories        lat        lng
0           Tandem Coffee     Coffee Shop  43.653559 -79.361809
1        Roselle Desserts          Bakery  43.653447 -79.362017
2  Figs Breakfast & Lunch  Breakfast Spot  43.655675 -79.364503
3         The Yoga Lounge     Yoga Studio  43.655515 -79.364955
4     Body Blitz Spa East             Spa  43.654735 -79.359874

<i> Number of venues  returned by Foursquare

In [23]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

21 venues were returned by Foursquare.


##### 3.5. Similar Exploration of all other Neighborhoods in out Toronto dataframe
###### 3.5.1 <i>  a function to repeat the same process  to all the neighborhoods in Toronto

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

###### <i> 3.5.2 Now  the code to run the above function on each neighborhood and create a new dataframe called Toronto_venues

In [25]:
Toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

###### <i> 3.5.3 checking the size of the resulting dataframe

In [26]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1514, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront                43.6555                -79.3626   
1  Regent Park, Harbourfront                43.6555                -79.3626   
2  Regent Park, Harbourfront                43.6555                -79.3626   
3  Regent Park, Harbourfront                43.6555                -79.3626   
4  Regent Park, Harbourfront                43.6555                -79.3626   

                    Venue  Venue Latitude  Venue Longitude  Venue Category  
0           Tandem Coffee       43.653559       -79.361809     Coffee Shop  
1        Roselle Desserts       43.653447       -79.362017          Bakery  
2  Figs Breakfast & Lunch       43.655675       -79.364503  Breakfast Spot  
3         The Yoga Lounge       43.655515       -79.364955     Yoga Studio  
4     Body Blitz Spa East       43.654735       -79.359874             Spa

1,514 venues were returned for the 39 neighborhoods of inhterest

###### <i> 3.5.4 checking how many venues were returned for each neighborhood

In [27]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            94   
Brockton, Parkdale Village, Exhibition Place                           39   
CN Tower, King and Spadina, Railway Lands, Harb...                     57   
Central Bay Street                                                     59   
Christie                                                               12   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             23   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           18   
Enclave of M4L                                                         14   
Enclave of M5E                                                         98   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                       4   
High Park, The Junction South                                           2   
India Bazaar, The Beaches West                                         22   
Kensington Market, Chinatown, Grange Park                              54   
Lawrence Park                                                           2   
Little Portugal, Trinity                                               47   
Moore Park, Summerhill East                                             7   
North Toronto West                                                      4   
Parkdale, Roncesvalles                                                 42   
Regent Park, Harbourfront                                              21   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     49   
St. James Town                                                         92   
St. James Town, Cabbagetown                                            45   
Studio District                                                         7   
Summerhill West, Rathnelly, South Hill, Forest ...                      6   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             9   
The Danforth  East                                                      7   
The Danforth West, Riverdale                                           35   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         27   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             94   
Brockton, Parkdale Village, Exhibition Place                            39   
CN Tower, King and Spadina, Railway Lands, Harb...                      57   
Central Bay Street                                                      59   
Christie                                                                12   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              23   
Davisville North                                                         8   
Dufferin, Dovercourt Village        

###### <i> 3.5.5 finding out how many unique categories can be curated from all the returned venues

In [28]:
print('There are {} unique categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 218 unique categories.


##### 3.6 Analyzing Each Neighborhood

In [29]:
# one hot encoding :  Creating a dataframe named "Toronto_onehot" whose columns are the didifferent  venue categories
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to "Toronto_onehot" dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
first_column = Toronto_onehot.pop('Neighborhood')
Toronto_onehot.insert(0, 'Neighborhood', first_column)


Toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  \
0  Regent Park, Harbourfront                  0               0   
1  Regent Park, Harbourfront                  0               0   
2  Regent Park, Harbourfront                  0               0   
3  Regent Park, Harbourfront                  0               0   
4  Regent Park, Harbourfront                  0               0   

   Afghan Restaurant  American Restaurant  Art Gallery  Art Museum  \
0                  0                    0            0           0   
1                  0                    0            0           0   
2                  0                    0            0           0   
3                  0                    0            0           0   
4                  0                    0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       1     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Stadium  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0          0         0           0                   0   
1                   0          0         0           0                   0   
2                   0          0         0           0                   0   
3                   0          0         0           0                   0   
4                   0          0         0           0                   0   

   Bistro  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0       0          0         0                     0               0        0   
1       0          0         0                     0               0        0   
2       0          0         0                     0               1        0   
3       0          0         0                     0               0        0   
4       0          0         0                     0               0        0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Bus Stop  \
0                0             0              0         0         0   
1                0             0              0         0         0   
2                0             0              0         0         0   
3                0             0              0         0         0   
4                0             0              0         0         0   

   Business Service  Butcher  Café  Camera Store  Candy Store  \
0                 0        0     0             0            0   
1                 0        0     0             0            0   
2                 0        0     0             0            0   
3                 0        0     0             0            0   
4                 0        0     0             0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chiropractor  \
0                     0            0                   0             0   
1                     0            0                   0             0   
2                     0            0                   0             0   
3                     0            0                   0             0   
4                     0            0                   0             0   

   Church  Clothing Store  Cocktail Bar  Coffee Shop  College Arts Building  \
0       0               0             0            1      

Above is, the number of occurences for each unique type of venue is displayed per neighborhood with neighborhood rows replicated depending on the different venues they host

###### 3.6.1 <i> examining the new dataframe size.

In [30]:
Toronto_onehot.shape

(1514, 218)

###### <i> 3.6.2  Next, grouping rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [31]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Accessories Store  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.025641   
2   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
3                                  Central Bay Street           0.000000   
4                                            Christie           0.000000   
5                                Church and Wellesley           0.000000   
6                      Commerce Court, Victoria Hotel           0.000000   
7                                          Davisville           0.000000   
8                                    Davisville North           0.000000   
9                        Dufferin, Dovercourt Village           0.000000   
10                                     Enclave of M4L           0.000000   
11                                     Enclave of M5E           0.000000   
12             First Canadian Place, Underground city           0.000000   
13                           Forest Hill North & West           0.000000   
14                           Garden District, Ryerson           0.000000   
15  Harbourfront East, Union Station, Toronto Islands           0.000000   
16                      High Park, The Junction South           0.000000   
17                     India Bazaar, The Beaches West           0.000000   
18          Kensington Market, Chinatown, Grange Park           0.000000   
19                                      Lawrence Park           0.000000   
20                           Little Portugal, Trinity           0.000000   
21                        Moore Park, Summerhill East           0.000000   
22                                 North Toronto West           0.000000   
23                             Parkdale, Roncesvalles           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                    Studio District           0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
33                The Annex, North Midtown, Yorkville           0.000000   
34                                        The Beaches           0.000000   
35                                 The Danforth  East           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    Adult Boutique  Afghan Restaurant  American Restaurant  Art Gallery  \
0         0.000000           0.000000             0.000000     0.021277   
1         0.000000           0.000000             0.000000     0.025641   
2         0.000000           0.000000             0.000000     0.000000   
3         0.000000           0.000000             0.000000     0.000000   
4         0.000000           0.000000             0.000000     0.000000   
5         0.012821           0.012821             0.012821     0.000000   
6         0.000000           0.000000             0.030000     0.010000   
7         0.000000           0.000000             0.000000     0.000000   
8         0.000000           0.000000             0.000000     0.000000   
9         0.000000           0.000000             0.000000     0.055556   
10        0.000000           0.000000             0.000000     0.000000   
11        0.000000           0.000000             0.000000     0.010204   
12        0.000000       

In [32]:
 # confirming the new size
Toronto_grouped.shape

(39, 218)

###### <i> 3.6.3  Let's print each neighborhood along with the top 5 most common venues


In [33]:
num_top_venues = 5

for hood in Toronto_grouped ['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.13
1               Hotel  0.05
2              Bakery  0.05
3        Cocktail Bar  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.08
1             Coffee Shop  0.08
2          Breakfast Spot  0.05
3  Thrift / Vintage Store  0.05
4                     Bar  0.05


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0  Italian Restaurant  0.07
1         Coffee Shop  0.07
2                Café  0.05
3                 Bar  0.05
4           Speakeasy  0.04


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.20
1      Sandwich Place  0.05
2                Café  0.05
3          Restaurant  0.03
4  Italian Restaurant  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  

###### <i> 3.7 Let's put that into a _pandas_ dataframe

<b> 3.7.1 First, let's write a function to sort the venues in descending order.

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<b> 3.7.2 Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [35]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...    Italian Restaurant   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                 Hotel                Bakery            Cocktail Bar   
1           Coffee Shop        Breakfast Spot  Thrift / Vintage Store   
2           Coffee Shop                  Café                     Bar   
3        Sandwich Place                  Café              Restaurant   
4                  Café            Playground             Coffee Shop   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0    Seafood Restaurant                  Café        Japanese Restaurant   
1                   Bar             Gift Shop         Mexican Restaurant   
2             Speakeasy                  Bank                       Park   
3    Italian Restaurant       Bubble Tea Shop  Middle Eastern Restaurant   
4            Baby Store                  Park         Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar            Restaurant               Pharmacy  
1              Boutique               Brewery           Liquor Store  
2     French Restaurant  Gym / Fitness Center          Grocery Store  
3     Electronics Store     French Restaurant         Breakfast Spot  
4           Candy Store   Moroccan Restaurant    Monument / Landmark

Above we displayed the ten types unique venues per neighborhood in order of popularity 

##### 3.8 Clustering Neighborhoods


###### <i> 3.8.1  Run k-means to cluster the neighborhoods into 5 clusters.

In [36]:
# set number of clusters
kclusters = 5

# dropping the "Neighborhood" column that is a string and not required in clustering
Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 4, 0, 0, 0, 0, 0], dtype=int32)

###### <i> 3.8.2 Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [37]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df_viz

# merge Toronto_grouped with Toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.tail() # check the last 5 columns!

PostalCode                 Borough                            Neighborhood  \
34        M5W  Downtown Toronto Stn A                          Enclave of M5E   
35        M4X        Downtown Toronto             St. James Town, Cabbagetown   
36        M5X        Downtown Toronto  First Canadian Place, Underground city   
37        M4Y        Downtown Toronto                    Church and Wellesley   
38        M7Y   East Toronto Business                          Enclave of M4L   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
34   43.6437   -79.3787               0           Coffee Shop   
35   43.6684   -79.3689               0           Coffee Shop   
36   43.6492   -79.3823               0           Coffee Shop   
37   43.6656   -79.3830               0   Japanese Restaurant   
38   43.7804   -79.2505               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
34            Restaurant                 Hotel                   Gym   
35            Restaurant                Bakery           Pizza Place   
36                 Hotel                  Café            Restaurant   
37           Coffee Shop      Sushi Restaurant               Gay Bar   
38            Restaurant   Japanese Restaurant           Salad Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
34         Deli / Bodega                  Café   Japanese Restaurant   
35    Italian Restaurant             Pet Store                  Café   
36                   Gym         Deli / Bodega           Salad Place   
37            Restaurant           Yoga Studio                   Gym   
38             Bookstore      Sushi Restaurant                  Bank   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
34    Sporting Goods Shop    Italian Restaurant                   Park  
35  General Entertainment                Market               Bus Stop  
36     Seafood Restaurant            Steakhouse    Japanese Restaurant  
37   Fast Food Restaurant          Burger Joint            Men's Store  
38         Breakfast Spot           Yoga Studio           Intersection

Above we show which cluster each of the neighborhoods fall into

###### <i> 3.9 Finally, let's visualize the resulting clusters

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)

# Display the clusters
map_clusters

The resulting clustering visual shows that most of the neighborhoods are very similar with the rest of the clusters comprising very few neighbourhoods each. 

<b><i><font color=red> if map is not showing you can try this nbviewer link to the same Github file </font> </i></b>  
https://nbviewer.jupyter.org/github/Karumaj/Coursera_IBMADS_Capstone/blob/master/ADSC_Toronto_Neighborhoods.ipynb


## 3.10. Examining Clusters
Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster..


###### <i> 3.10.1 Cluster 1

In [39]:
cluster_1 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]
print(cluster_1.shape)
cluster_1

(32, 12)


Borough  Cluster Labels        1st Most Common Venue  \
0         Downtown Toronto               0                  Coffee Shop   
1         Downtown Toronto               0                  Coffee Shop   
2         Downtown Toronto               0                  Coffee Shop   
3             East Toronto               0                          Pub   
4         Downtown Toronto               0                  Coffee Shop   
5         Downtown Toronto               0                  Coffee Shop   
7         Downtown Toronto               0                         Café   
8             West Toronto               0                Grocery Store   
9   East York/East Toronto               0                         Park   
10        Downtown Toronto               0                  Music Venue   
11            West Toronto               0                          Bar   
12            East Toronto               0             Greek Restaurant   
13        Downtown Toronto               0                  Coffee Shop   
14            West Toronto               0                         Café   
15            East Toronto               0                         Park   
16        Downtown Toronto               0                  Coffee Shop   
17            East Toronto               0              Coworking Space   
20         Central Toronto               0                   Playground   
21         Central Toronto               0             Business Service   
24         Central Toronto               0               Sandwich Place   
25            West Toronto               0  Eastern European Restaurant   
26         Central Toronto               0               Sandwich Place   
27        Downtown Toronto               0                         Café   
28            West Toronto               0                         Café   
30        Downtown Toronto               0                         Café   
31         Central Toronto               0           Light Rail Station   
32        Downtown Toronto               0           Italian Restaurant   
34  Downtown Toronto Stn A               0                  Coffee Shop   
35        Downtown Toronto               0                  Coffee Shop   
36        Downtown Toronto               0                  Coffee Shop   
37        Downtown Toronto               0          Japanese Restaurant   
38   East Toronto Business               0                  Coffee Shop   

            2nd Most Common Venue 3rd Most Common Venue  \
0                  Breakfast Spot           Yoga Studio   
1                  Clothing Store                  Café   
2                            Café    Seafood Restaurant   
3               Health Food Store                 Trail   
4                           Hotel                Bakery   
5                  Sandwich Place                  Café   
7                     Coffee Shop            Restaurant   
8                          Bakery                  Park   
9                     Coffee Shop          Intersection   
10                Harbor / Marina                  Café   
11                    Coffee Shop          Cocktail Bar   
12                 Ice Cream Shop                  Café   
13                          Hotel                  Café   
14                    Coffee Shop        Breakfast Spot   
15                     Restaurant  Fast Food Restaurant   
16                          Hotel                  Café   
17                          Diner        Baseball Field   
20           Gym / Fitness Center        Breakfast Spot   
21     Construction & Landscaping                  Park   
24                           Café     French Restaurant   
25                    Coffee Shop                  Café   
26                   Dessert Shop           Coffee Shop   
27            Japanese Restaurant             Bookstore   
28                    Pizza Place                   Pub   
30  Vegetarian / Vegan Restaurant           Coffee Shop   
31       

###### <i> 3.10.2 Cluster 2

In [40]:
cluster_2 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]
print(cluster_2.shape)
cluster_2

(1, 12)


Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               1          Home Service   

   2nd Most Common Venue    3rd Most Common Venue 4th Most Common Venue  \
19     Accessories Store  New American Restaurant    Mexican Restaurant   

        5th Most Common Venue 6th Most Common Venue  \
19  Middle Eastern Restaurant    Miscellaneous Shop   

         7th Most Common Venue            8th Most Common Venue  \
19  Modern European Restaurant  Molecular Gastronomy Restaurant   

   9th Most Common Venue 10th Most Common Venue  
19   Monument / Landmark    Moroccan Restaurant

###### <i> 3.10.3 Cluster 3

In [41]:
cluster_3 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]
print(cluster_3.shape)
cluster_3

(1, 12)


Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               2    Photography Studio   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                  Park     Accessories Store           Music Store   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
18           Men's Store    Mexican Restaurant  Middle Eastern Restaurant   

   8th Most Common Venue       9th Most Common Venue  \
18    Miscellaneous Shop  Modern European Restaurant   

             10th Most Common Venue  
18  Molecular Gastronomy Restaurant

In [42]:
cluster_4 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]
print(cluster_4.shape)
cluster_4

(1, 12)


Borough  Cluster Labels 1st Most Common Venue  \
22  West Toronto               3                  Park   

                       2nd Most Common Venue 3rd Most Common Venue  \
22  Residential Building (Apartment / Condo)     Accessories Store   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
22           Music Store           Men's Store    Mexican Restaurant   

        7th Most Common Venue 8th Most Common Venue  \
22  Middle Eastern Restaurant    Miscellaneous Shop   

         9th Most Common Venue           10th Most Common Venue  
22  Modern European Restaurant  Molecular Gastronomy Restaurant

###### <i> 3.10.5 Cluster 5

In [43]:
cluster_5 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]
print(cluster_5.shape)
cluster_5

(4, 12)


Borough  Cluster Labels 1st Most Common Venue  \
6   Downtown Toronto               4         Grocery Store   
23   Central Toronto               4            Playground   
29   Central Toronto               4                  Park   
33  Downtown Toronto               4            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Café            Playground           Coffee Shop   
23              Gym Pool                  Park                Garden   
29            Playground       Thai Restaurant                   Gym   
33         Grocery Store                  Park           Candy Store   

   5th Most Common Venue 6th Most Common Venue  \
6             Baby Store                  Park   
23           Music Store           Men's Store   
29         Grocery Store                 Trail   
33           Music Venue    Mexican Restaurant   

              7th Most Common Venue      8th Most Common Venue  \
6                Athletics & Sports                Candy Store   
23               Mexican Restaurant  Middle Eastern Restaurant   
29  Molecular Gastronomy Restaurant              Movie Theater   
33        Middle Eastern Restaurant         Miscellaneous Shop   

         9th Most Common Venue           10th Most Common Venue  
6          Moroccan Restaurant              Monument / Landmark  
23          Miscellaneous Shop       Modern European Restaurant  
29         Moroccan Restaurant              Monument / Landmark  
33  Modern European Restaurant  Molecular Gastronomy Restaurant

#### 3.11 CLUSTERING RESULTS
Some general observations can be made about the uniqueness of the five resulting clusters as follows
##### cluster_1
<i> Above is the cluster with most neighborhoods - 32 out of 39 neighborhoods. From the abobe map, the neighbourborhoods occupy the coastline and are more concentrated on the central region of the city and are from all the boroughs in our exploration.

With cafes and eating places appering in most popular positions, they definitely represent the most urban places where most of the working population are found.</i>

##### cluster_2,3, & 4
<i> The three clusters are is only defined by one neighborhood each.
   They are in the outlying region of the city.
   They are probably surburban areas with charasteristic venues such as home service, accessories, studios and parks being most popular. </i>
   
##### cluster_5
Cluster 5 exhibits very common characterics among the neigborhoods that could be categorized as peri-urban neighborhoods  in relation to the other four since geographically from the map visual they appear spread in the areas between the core city neighborhoods and the outlying regions. They are probably mainly residential areas due to the common popularity of playgrounds, gym and grocery venues. 